time/timesyncd.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	"github.com/godbus/dbus/v5/prop"
+	"github.com/home-assistant/os-agent/utils/lineinfile"
+	"regexp"
+	"strings"
+
+	logging "github.com/home-assistant/os-agent/utils/log"
+)
+
+const (
+	objectPath    = "/io/hass/os/Time/Timesyncd"
+	ifaceName     = "io.hass.os.Time.Timesyncd"
+	timesyncdConf = "/etc/systemd/timesyncd.conf"
+)
+
+var (
+	optNTPServer         []string
+	optFallbackNTPServer []string
+	configFile           = lineinfile.LineInFile{FilePath: timesyncdConf}
+)
+
+type timesyncd struct {
+	conn  *dbus.Conn
+	props *prop.Properties
+}
+
+func getNTPServers() []string {
+	return getTimesyncdConfigProperty("NTP")
+}
+
+func getFallbackNTPServers() []string {
+	return getTimesyncdConfigProperty("FallbackNTP")
+}
+
+func setNTPServer(c *prop.Change) *dbus.Error {
+	servers, ok := c.Value.([]string)
+	if !ok {
+		return dbus.MakeFailedError(fmt.Errorf("invalid type for NTPServer"))
+	}
+
+	value := strings.Join(servers, " ")
+
+	if err := setTimesyncdConfigProperty("NTP", value); err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	optNTPServer = servers
+	return nil
+}
+
+func setFallbackNTPServer(c *prop.Change) *dbus.Error {
+	servers, ok := c.Value.([]string)
+	if !ok {
+		return dbus.MakeFailedError(fmt.Errorf("invalid type for FallbackNTPServer"))
+	}
+
+	value := strings.Join(servers, " ")
+
+	if err := setTimesyncdConfigProperty("FallbackNTP", value); err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	optFallbackNTPServer = servers
+	return nil
+}
+
+func getTimesyncdConfigProperty(property string) []string {
+	value, err := configFile.Find(`^\s*(`+property+`=).*$`, `\[Time\]`, true)
+
+	var servers []string
+
+	if err != nil || value == nil {
+		return servers
+	}
+
+	matches := regexp.MustCompile(property + `=([^\s#]+(?:\s+[^\s#]+)*)`).FindStringSubmatch(*value)
+	if len(matches) > 1 {
+		servers = strings.Split(matches[1], " ")
+	}
+
+	return servers
+}
+
+func setTimesyncdConfigProperty(property string, value string) error {
+	var params = lineinfile.NewPresentParams("NTP=" + value)
+	params.Regexp, _ = regexp.Compile(`^\s*#?\s*(` + property + `=).*$`)
+	// Keep it simple, timesyncd.conf only has the [Time] section
+	params.After = `\[Time\]`
+	if err := configFile.Present(params); err != nil {
+		return fmt.Errorf("failed to set %s: %w", property, err)
+	}
+
+	if err := restartTimesyncd(); err != nil {
+		return fmt.Errorf("failed to restart timesyncd: %w", err)
+	}
+
+	return nil
+}
+
+func restartTimesyncd() error {
+	conn, err := dbus.SystemBus()
+	if err != nil {
+		return err
+	}
+
+	obj := conn.Object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+	call := obj.Call("org.freedesktop.systemd1.Manager.RestartUnit", 0, "systemd-timesyncd.service", "replace")
+	if call.Err != nil {
+		return call.Err
+	}
+
+	return nil
+}
+
+func InitializeDBus(conn *dbus.Conn) {
+	d := timesyncd{
+		conn: conn,
+	}
+
+	optNTPServer = getNTPServers()
```

Comment:
I guess this could be a bit racy, depending on when OS Agent/NetworkManager hook gets called :thinking:

Reply:
The current implementation only touches the persistent configuration file, so NetworkManager hook doesn't affect it in any way. If we want to read the dynamically obtained NTPs written by the NM hook, maybe using a different method/property would be better.

Reply:
Why do we even need these getters? There is `FallbackNTPServers` and `SystemNTPServers` available on the `/org/freedesktop/timesync1` object of the `org.freedesktop.timesync1` service.

In a way, this interface to me is mainly a stopgap until systemd-timesyncd has a similar functionality.

Reply:
Those can be amalgamation of multiple configuration files as described in [man timesyncd.conf](https://man7.org/linux/man-pages/man5/timesyncd.conf.5.html). This interface exposes what is statically configured in the file we'd like to be the source of truth for the configuration. I can imagine the "native" Systemd attributes could be used for sanity checking, raising an issue if the configured value doesn't match what's in the config (which e.g. means user has created extra configuration files). But it would be confusing if you had only the setters, and under some circumstances the getters would return different value than what you have set.

I also doubt that Systemd will ever get similar functionality. There are the [runtime NTP servers](https://www.github.com/systemd/systemd/pull/22614) but those we can't use because we need to apply the configuration before our "configuration daemon" (Supervisor) is started. AFAIK Systemd doesn't have a way to edit configuration files through D-Bus API anywhere, and some things can be only configured only through config files.

Reply:
> I think systemd sees NTP server configuration a network manager thing, just like DNS. And it should be handled and stored by the network manager, and transferred at the appropriate time (the appropriate link is actually ready). I mean, you can imagine a system where you have two network links, each have their own local NTP servers.

Yes, but I think this is not achievable without using systemd-networkd at the same time. Currently we store the network config in NetworkManager's config files, where we lack any mechanism to save per-link user-configured NTP servers, so we'd have to introduce one (i.e. have our own config files/attributes of individual NM connections that'd somehow propagate to timesyncd config). But the complexity of such implementation would heavily outweigh any maintenance cost of the current approach to this problem.

Anyway, if we shall consider multihomed setups which even require different NTP servers on individual links, then I think this discussion turns academic. AFAIK most of the multihomed setups are not what HA considers _anywhere_, leave aside the fact that even just the NTP config alone seems _too advanced_ for some :sweat_smile: We really should think about the real use case, and the way we want to handle NTP configuration on user's side.

I agree though that we can take a different approach on this problem and instead of exposing "time subsystem" APIs, we could have "system configuration" instead, if that makes more sense to you.